pkg/command/plugin_bundle.go

```diff
@@ -249,3 +261,110 @@ func completeGroupsAndVersionForBundleDownload(_ *cobra.Command, _ []string, toC
+	comps := make([]string, len(versions))
+	for i := range versions {
+		comps[len(versions)-1-i] = fmt.Sprintf("%s:%s", id, versions[i])
+	}
+	return comps
+}
+
+func completionGetPluginsForBundleDownload() ([]*plugininventory.PluginInventoryEntry, error) {
+	// For a download-bundle, we cannot use the DB cache.  This is because
+	// the download-bundle does not use the configured plugin sources.  Instead it
+	// uses the repo specified by the `--image`` flag, or, without the `--image` flag,
+	// it uses the default central repo automatically.
+
+	// We start by downloading the inventory of the required repo.  This is not
+	// very fast, but there isn't much we can do about it.
+
+	var err error
+	tempDBDir, err := os.MkdirTemp("", "")
+	if err != nil {
+		return nil, err
+	}
+	defer os.RemoveAll(tempDBDir)
+
+	// Download the plugin inventory oci image to tempDBDir
+	repoImage := dpbo.pluginDiscoveryOCIImage
+	inventoryFile := filepath.Join(tempDBDir, plugininventory.SQliteDBFileName)
+	if err := imageProcessorForDownloadBundleComp.DownloadImageAndSaveFilesToDir(repoImage, filepath.Dir(inventoryFile)); err != nil {
+		return nil, err
+	}
+
+	// Read the plugin inventory database to read the plugin groups it contains
+	pi := plugininventory.NewSQLiteInventory(inventoryFile, path.Dir(repoImage))
+	pluginEntries, err := pi.GetPlugins(&plugininventory.PluginInventoryFilter{IncludeHidden: true}) // Include the hidden plugin groups during plugin migration
+	if err != nil {
+		return nil, err
+	}
+	return pluginEntries, err
+}
+
+func completePluginIDForBundleDownload(_ *cobra.Command, _ []string, toComplete string) ([]string, cobra.ShellCompDirective) {
+	PluginEntries, err := completionGetPluginsForBundleDownload()
+	if err != nil {
+		return nil, cobra.ShellCompDirectiveNoFileComp
+	}
+	if PluginEntries == nil {
+		comps := cobra.AppendActiveHelp(nil, fmt.Sprintf("There are no plugins in %s", dpbo.pluginDiscoveryOCIImage))
+		return comps, cobra.ShellCompDirectiveNoFileComp
+	}
+
+	if idx := strings.Index(toComplete, ":"); idx != -1 {
+		// The pluginID is already specified before the :
+		// so now we should complete the group version.
+		// Since more recent versions are more likely to be
+		// useful, we tell the shell to preserve the order
+		// using cobra.ShellCompDirectiveKeepOrder
+		pluginID := toComplete[:idx]
+		versionToComplete := toComplete[idx+1:]
+
+		return completeVersionInPluginIDForDownloadBundle(PluginEntries, pluginID, versionToComplete), cobra.ShellCompDirectiveNoFileComp | cobra.ShellCompDirectiveKeepOrder
+	}
+
+	// Complete plugin group names
+	var comps []string
+	for _, p := range PluginEntries {
+		id := plugininventory.PluginToID(p)
+		if strings.HasPrefix(id, toComplete) {
+			comps = append(comps, id)
+		}
+	}
+
+	if len(comps) == 0 {
+		return cobra.AppendActiveHelp(nil, fmt.Sprintf("There are no plugins matching: '%s'", toComplete)), cobra.ShellCompDirectiveNoFileComp | cobra.ShellCompDirectiveKeepOrder
```

Comment:
Although this is a good idea, it will no longer work if we stop using `HasPrefix()` because our go code doesn't know in advance if the shell will match on anything or not.

This is a limitation of supporting fuzzy matching.